### ISSUE TITLE
Flatpak/flathub support

### ISSUE THREAD

**Issue by TeaDrinkingProgrammer** [OWASP/threat-dragon#626]
**Describe what problem your feature request solves**
A Flatpak install would be a great alternative to Snaps and Appimages, since this allows most users to download threat-dragon straight from their software store. Would people be open for this? I could look into it myself but I want to see if there is an appetite for it.

**Comment by jgadsden** [comment]
This would be great to have @TeaDrinkingProgrammer , if you would like to find out what is involved then that would be great

**Comment by jgadsden** [comment]
We probably could follow the example in https://docs.flatpak.org/en/latest/electron.html
it takes several steps, but nothing that can not be automated

**Comment by TeaDrinkingProgrammer** [comment]
@sparticvs @jgadsden  I was still planning on looking at it, it is just that I am incredibly busy right now, I probably will have some time in a week or two. I have never done anything with Flatpak though, so it will be a learning experience for me. How about I start working on it and ask for help/feedback when I have something to show for?

**Comment by jgadsden** [comment]
Yes, sure, that sounds great @TeaDrinkingProgrammer  - over to you

**Comment by TeaDrinkingProgrammer** [comment]
I did a bit of research and bumped into this fun roadblock:
The guide @jgadsden linked is indeed very useful and I followed it. This went well until I got to [the step where you have to build the NPM packages](https://docs.flatpak.org/en/latest/electron.html#bundling-npm-packages). This step uses a separate python script to download all NPM dependencies so you do not have to put them in the manifest manually. The problem is that this script relies on a package-lock file from either NPM or YARN to get the download links. I have looked up if maybe it would be possible to convert from PNPM lock files to NPM lock files, [but the creator of PNPM has the staunch opinion developers should add support for PNPM instead of PNPM adding a convertor](https://github.com/orgs/pnpm/discussions/3367#discussioncomment-4214798) (I think this is quite unrealistic to expect from users and that it will put off developers rather than help adoption, but I digress).

I think there are a few ways of solving this issue:
- Switching to NPM: I do not know the ins and outs of PNPM vs NPM, but I imagine it was chosen for a reason.
- Adding PNPM support to the script: this would be possible, the script is not very big, but I personally wouldn't want to undertake it as I am not very fluent in Python, Flatpak or PNPM.
- Only using NPM for generating a package-lock: I have tried this out. Besides the fact that the only-allow package would need to be disabled for this, it also creates a package-lock with different versions than the main one, which will probably create some headaches.

I honestly do not like any of the options, but I do not see any others. Any ideas?

**Comment by jgadsden** [comment]
Hello @TeaDrinkingProgrammer  - I realise the amount of work to get this far, much appreciated

pnpm was chosen because it is fast on the download and install ... and we liked using it as it is marginally more forgiving than npm.  But that was the only reason: we thought it looked fun and went with it

It has caused us some problems, for example moving from pnpm version 7 to 8, and it does have some downsides

If it is stopping us from doing flatpak then we can migrate back to npm, it is not as bad as it sounds and does simplify our pipeline scripts

We can ponder for a bit? and if we decide to move back to npm then I am happy to implement this, unless you want the honour :)

**Comment by jgadsden** [comment]
I was intending to do a 2.1 release soon, this will reinstate the PDF button on the desktop version

I could delay this for a bit so as to get npm back in if we decide that is the way forward ... all part of the pondering :)

**Comment by TeaDrinkingProgrammer** [comment]
@jgadsden No, I'm happy to hand it back over to you, it is a lot more complicated than I thought 😅 , but hey, I learned a lot from it. Probably, moving to NPM is the obvious option as it is the industry default, but the script also supports YARN so that would also be an option. From what I can find, Yarn is faster than NPM, but slower or equal to PNPM, but I find that these kinds of benchmarks or opinions are rather subjective. One interesting thing about Yarn, is that it does not have an audit fix command [(you need a package for this)](https://www.npmjs.com/package/yarn-audit-fix), I can imagine this could be a dealbreaker for an app under the OWASP umbrella. I would personally give them both a try and see what works, but that's just me.

**Comment by jgadsden** [comment]
oops, I did not mean to say that I wanted to implement the flatpak, I was just referring to the migration back to npm from pnpm
I hope you will still carry on with the flatpak @TeaDrinkingProgrammer ?

**Comment by TeaDrinkingProgrammer** [comment]
I have heard about Fedora turning RPMs into flatpaks, would that be worth looking in to?

**Comment by sparticvs** [comment]
@TeaDrinkingProgrammer yes, although I haven't done that myself, I know others that have.

**Comment by jgadsden** [comment]
We have moved Threat Dragon back to npm from pnpm. There is no real difference, npm install times now seem to be comparable with pnpm and we are no longer forcing our developers to use pnpm.
It is still OK to use pnpm with Threat Dragon, we do not get in the way of anyone installing and running with pnpm, it is just that now the repo clones with a `package-lock.json` instead of a` pnpm-lock.json` file

@TeaDrinkingProgrammer did you want to carry on with your development? you probably just need to sync your fork to get the npm lock file

**Comment by TeaDrinkingProgrammer** [comment]
It's been a while, but I haven't forgotten about this. I have looked at it again with NPM and I have gotten a lot further, but I am up to a roadblock again. I have a strong hunch it has to do with the Python script but I will discuss it here first to rule out any oddities on this side.

The problem is as follows:
As I discussed previously, Flatpak does not allow internet connections in the build step. Flatpak does support downloading sources before the build step, so there is a script which "converts" the package-lock.json file to a generated-sources.json file. This allows flatpak to download all node_modules dependencies before  the buildstep which are then used by NPM during npm install. I generate the generated-sources file first and let Flatpack build, but get this error:

**Comment by jgadsden** [comment]
I will try these commands on my system, ND so how far I get

**Comment by mademarc** [comment]
Hey @jgadsden can i contribute to this issue?  and if so how can i help & contribute to get tis resolve ASAP.

Please and thanks :)

**Comment by jgadsden** [comment]
Hello @mademarc , if you would like to repeat what @TeaDrinkingProgrammer has done so far, that would be a great first step

**Comment by jgadsden** [comment]
@sparticvs , @TeaDrinkingProgrammer and @mademarc , is the flatpak / flathub support progressing? We moved to npm and `package-lock.json` (instead of a `pnpm-lock.json` file) so it may be worth revisiting
No problem if we can not do Flatpak in time for version 2.3 (end August 2024) as it can be bumped back to version 2.4 (early December 2024)

**Comment by TeaDrinkingProgrammer** [comment]
I might look into it in a couple of weeks so if I find the time, it would be in time for 2.3.

**Comment by jgadsden** [comment]
If I understand correctly, there is no need for a flathub account? The submission is done following page [Submission](https://docs.flathub.org/docs/for-app-authors/submission/)

**Comment by mademarc** [comment]
hey @jgadsden i was  unable to address the Flatpak feature request asi was super busy at that moment and that said i could consider the following steps to assist that are below: 

1. Research Flatpak Packaging: Look into creating a Flatpak package, especially for Electron apps.

2.Test Locally: Build and test the Flatpak package on their system.

3. Submit to Flathub: Follow Flathub’s submission guidelines to make it available.

4. Update Documentation: Add Flatpak installation instructions to the project's documentation.

or can i follow up with @TeaDrinkingProgrammer in order to fix this issue ?

**Comment by jgadsden** [comment]
you are most welcome to work on this @mademarc , I do not have the time myself but if you could coordinate with @TeaDrinkingProgrammer that would be great
